Sources/OpenTelemetrySdk/Trace/TracerProviderSdk.swift

```diff
@@ -158,3 +144,63 @@ public class TracerProviderSdk: TracerProvider {
     sharedState.activeSpanProcessor.forceFlush(timeout: timeout)
   }
 }
+
+// Stolen from openTelemtryApi.
+private final class TraceProviderLock {
+    private let mutex: UnsafeMutablePointer<pthread_mutex_t> = UnsafeMutablePointer.allocate(capacity: 1)
+    
+    /// Create a new lock.
+    init() {
+        let err = pthread_mutex_init(mutex, nil)
+        precondition(err == 0, "pthread_mutex_init failed with error \(err)")
+    }
+    
+    deinit {
+        let err = pthread_mutex_destroy(self.mutex)
+        precondition(err == 0, "pthread_mutex_destroy failed with error \(err)")
+        self.mutex.deallocate()
+    }
+    
+    /// Acquire the lock.
+    ///
+    /// Whenever possible, consider using `withLock` instead of this method and
+    /// `unlock`, to simplify lock handling.
+    func lock() {
+        let err = pthread_mutex_lock(mutex)
+        precondition(err == 0, "pthread_mutex_lock failed with error \(err)")
+    }
+    
+    /// Release the lock.
+    ///
+    /// Whenever possible, consider using `withLock` instead of this method and
+    /// `lock`, to simplify lock handling.
+    func unlock() {
+        let err = pthread_mutex_unlock(mutex)
+        precondition(err == 0, "pthread_mutex_unlock failed with error \(err)")
+    }
+}
+
+extension TraceProviderLock {
+    /// Acquire the lock for the duration of the given block.
+    ///
+    /// This convenience method should be preferred to `lock` and `unlock` in
+    /// most situations, as it ensures that the lock will be released regardless
+    /// of how `body` exits.
+    ///
+    /// - Parameter body: The block to execute while holding the lock.
+    /// - Returns: The value returned by the block.
+    @inlinable
+    func withLock<T>(_ body: () throws -> T) rethrows -> T {
+        lock()
+        defer {
+            self.unlock()
+        }
+        return try body()
+    }
+    
+    // specialise Void return (for performance)
+    @inlinable
+    func withLockVoid(_ body: () throws -> Void) rethrows {
+        try withLock(body)
+    }
+}
```

Comment:
looks good to me, please run this through swift-format, looks like you're using 4-space tabs, instead of 2.